GpsDataAnalyzer/calculator/alignment_algorithms.py

```diff
@@ -0,0 +1,293 @@
+        if rounded_time in time_points_mapping:
+            time_points_mapping[rounded_time].append(point)
+        else:
+            time_points_mapping[rounded_time] = [point]
+    return time_points_mapping
+
+def find_optimal_offset(primary_set, secondary_set, start_time, 
+                        offset_range_length, point_checking_range_length):
+    """
+    Find optimal offset from offset range to shift the secondary set
+
+    Finds the optimal offset meaning, an offset that is in the given offset
+    range to check and that also has overlap between the two sets. If the
+    mean distance is over 100m, it returns None, since this indicates sets
+    that do not align even if their timestamps can be shifted to align. 
+
+    Args:
+        primary_set: dictionary in format {DateTime: [GpsData,], ...}
+        secondary_set: dictionary in format {DateTime: [GpsData,], ...}
+        start_time: Datetime to start calculation at
+        offset_range: int, how many offset values to check i.e. the range of 
+                      (-offset_range//2, offset_range//2)
+        point_checking_range: int, how many points around the start to use in the deviation
+                              mean calculations
+        primary_set_index: which set (1 or 2) is the primary (i.e. not being shifted)
+
+    Returns:
+        int, the offset in seconds that should be applied to the secondary
+        data set if an optimal offset exists,
+        None otherwise
+    """
+    optimal_offset= 0
+    optimal_mean_distance = None
+
+    for offset in range (-offset_range_length//2, offset_range_length//2):
+        distances = []
+        skipped_point_count = 0
+        for i in range(-point_checking_range_length//2, point_checking_range_length//2):
+            # apply offset to secondary set
+            primary_time = start_time + timedelta(seconds=i)
+            secondary_time = start_time + timedelta(seconds=i+offset)
+
+            if primary_time in primary_set and secondary_time in secondary_set:
+                # if more than one point at that time, choose first one
+                # TODO(ameles) consider if more complex downsampling needed
+                primary_point = primary_set[primary_time][0]
+                secondary_point = secondary_set[secondary_time][0]
+                location1 = (primary_point.latitude, primary_point.longitude)
+                location2 = (secondary_point.latitude, secondary_point.longitude)
+                distances.append(utils.calculate_distance(location1, location2))
+            else:
+                skipped_point_count += 1
+
+        if distances and (optimal_mean_distance is None or np.mean(distances) < optimal_mean_distance):
+            optimal_mean_distance = np.mean(distances)
+            optimal_offset = offset
+
+    print("optimal mean distance: " + str(optimal_mean_distance))
+    if optimal_mean_distance is None or optimal_mean_distance > 100:
```

Comment:
create a constant for 100: MEAN_DISTANCE_THRESHOLD = 100. 

you can also assign the initial value of optimal_mean_distance = MEAN_DISTANCE_THRESHOLD +1 on L265, and remove 'optimal_mean_distance is None from L286.